### PR TITLE
Optimize DeepSeek-V4 MUSA execution paths

### PR DESCRIPTION
**Pull request by Joey-gvwal**
## Summary                                                                                                                                                                                       
   
  - Optimize DeepSeek-V4 MUSA execution paths for vLLM v0.22, covering sparse-attn indexer, prefill top-k/blockselect, materialized top-k ordering, and speculative metadata upload paths.         
  - Add TP8 prefill combo kernels and direct materialized top-k support to reduce prefill/indexer overhead.
  - Optimize DeepSeek-V4 MoE DeepGEMM prefill and fused MoE chunked execution on MUSA.
  - Add MUSA custom op bindings and kernel updates for DeepSeek-V4 cache store / indexer top-k paths.
  - Fix MUSA graph check behavior when compilation mode is unset.
  - Add regression/source tests for fused MoE chunking, indexer patch coverage, materialized top-k routing, and DeepSeek-V4 MUSA platform behavior.

## Test
 Run vllm serve and benchmark passed

```
# env
export SAFETENSORS_FAST_GPU=1
export VLLM_DEEP_GEMM_WARMUP=skip
export VLLM_MEMORY_PROFILER_ESTIMATE_CUDAGRAPHS=0
export VLLM_USE_DEEP_GEMM=1
export VLLM_USE_DEEP_GEMM_E8M0=0
export VLLM_WORKER_MULTIPROC_METHOD=spawn
export PYTHONUNBUFFERED=1
export VLLM_MUSA_DEEPSEEK_V4_TP8_PROFILE=aggressive_long_prefill
export VLLM_MUSA_DEEPSEEK_V4_DISABLE_AUX_OVERLAP=1
export VLLM_MUSA_DEEPSEEK_V4_MHC_PRE_DECODE_PRENORM_IMPL=deepgemm
export VLLM_MUSA_SPEC_METADATA_SINGLE_REQ_CACHE=1
# start vllm
vllm serve /home/dist/DeepSeek-V4-Flash-Base \
  --served-model-name deepseek \
  --gpu-memory-utilization 0.95 \
  --max-model-len 6144 \
  --max-num-seqs 1 \
  --max-num-batched-tokens 8195 \
  --tensor-parallel-size 8 \
  -ac.backend FLASHMLA \
  --port 25606 \
  --host 0.0.0.0 \
  --kv-cache-dtype fp8 \
  --no-enable-prefix-caching \
  --async-scheduling \
  --compilation-config '{"mode":"NONE","cudagraph_mode":"FULL_DECODE_ONLY","max_cudagraph_capture_size":4,"cudagraph_capture_sizes":[4],"cudagraph_copy_inputs":false}' \
  --speculative-config '{"method":"mtp","num_speculative_tokens":3}'
# benchmark
vllm bench serve \
  --backend openai-chat \
  --dataset-name random \
  --random-input-len 4096 \
  --random-output-len 1024 \
  --ignore-eos \
  --num-prompts 1 \
  --num-warmups 5 \
  --request-rate 10 \
  --max-concurrency 1 \
  --model deepseek \
  --tokenizer /home/dist/DeepSeek-V4-Flash-Base \
  --temperature 0 \
  --base-url http://127.0.0.1:25606 \
  --endpoint /v1/chat/completions
```

### CODE REVIEW

File: vllm_musa/patches/series/0013-MUSA-vllm.model_executor.layers.sparse_attn_indexer.patch [hunk 1]
```diff
@@ -255,6 +305,221 @@ index 9597708b6..230ed6d4b 100644
...
++    if total_seq_lens <= 0 or overselect <= 0:
++        return True
++
++    block_size = int(kv_cache.shape[1])
++    if block_size <= 0:
++        return False
++
++    row_starts = chunk.cu_seqlen_ks[:rows].clamp(min=0, max=total_seq_lens)
++    row_ends = chunk.cu_seqlen_ke[:rows].clamp(min=0, max=total_seq_lens)
++    context_lens = row_ends.reshape(rows, 1).contiguous()
```
[the remark this answers is not in the document]
Comment (Joey-gvwal) [reply vllm_musa/patches/series/0013-MUSA-vllm.model_executor.layers.sparse_attn_indexer.patch:350]: Using row_ends - row_starts would misalign logits columns for a non-zero start because the provider has no start-offset argument, so no need to fix.
